pkg/auth/server/tokenrequest/endpoints.go

```diff
@@ -21,31 +24,56 @@ const (
 )
 
 type endpointDetails struct {
-	publicMasterURL   string
-	originOAuthClient *osincli.Client
+	publicMasterURL     string
+	osinOAuthClient     *osincli.Client
+	osinOAuthClientFunc func() (*osincli.Client, error)
+	ready               chan struct{}
+	closeOnce           sync.Once
 }
 
 type Endpoints interface {
 	Install(mux login.Mux, paths ...string)
 }
 
-func NewEndpoints(publicMasterURL string, originOAuthClient *osincli.Client) Endpoints {
-	return &endpointDetails{publicMasterURL, originOAuthClient}
+func NewEndpoints(publicMasterURL string, osinOAuthClientFunc func() (*osincli.Client, error)) Endpoints {
+	return &endpointDetails{publicMasterURL: publicMasterURL, osinOAuthClientFunc: osinOAuthClientFunc, ready: make(chan struct{})}
 }
 
 // Install registers the request token endpoints into a mux. It is expected that the
 // provided prefix will serve all operations
 func (endpoints *endpointDetails) Install(mux login.Mux, paths ...string) {
 	for _, prefix := range paths {
-		mux.HandleFunc(path.Join(prefix, RequestTokenEndpoint), endpoints.requestToken)
-		mux.HandleFunc(path.Join(prefix, DisplayTokenEndpoint), endpoints.displayToken)
+		mux.HandleFunc(path.Join(prefix, RequestTokenEndpoint), endpoints.readyHandler(endpoints.requestToken))
+		mux.HandleFunc(path.Join(prefix, DisplayTokenEndpoint), endpoints.readyHandler(endpoints.displayToken))
 		mux.HandleFunc(path.Join(prefix, ImplicitTokenEndpoint), endpoints.implicitToken)
 	}
+
+	go wait.Until(func() {
```

Comment:
If you do it like this, you need to have a stop channel.

Why not simply make a getter for thing which returns osinOauthClient or returns an error.  I'm fine with failing requests to this endpoint until this bit is ready.